File: AppBox/ViewController/DropboxViewController/DropboxViewController.m
//
//  DropboxViewController.m
//  AppBox
//
//  Created by Vineet Choudhary on 28/11/16.
//  Copyright © 2016 Developer Insider. All rights reserved.
//

#import "DropboxViewController.h"


@implementation DropboxViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    //Log Screen
    [EventTracker logScreen:@"Dropbox Login"];
    
    //DB Authentication Notification
    [[NSNotificationCenter defaultCenter] addObserver:self selector:@selector(handleLoggedInNotification:) name:abDropBoxLoggedInNotification object:nil];
	
	//Set quit button title
	[buttonQuit setTitle:[UserData isLoggedIn] ? @"Cancel" : @"Quit"];
}

- (IBAction)buttonConnectDropboxTapped:(NSButton *)sender {
    [EventTracker logEventWithType:LogEventTypeAuthDropbox];
    
	//Authenticate user
	[DBClientsManager authorizeFromControllerDesktopV2:[NSWorkspace sharedWorkspace] controller:self loadingStatusDelegate:nil openURL:^(NSURL * _Nonnull url) {
		[[NSWorkspace sharedWorkspace] openURL:url];
	} scopeRequest:nil];
}

- (IBAction)buttonQuitTapped:(NSButton *)sender {
    [EventTracker logEventWithType:LogEventTypeExitWithoutAuth];
    [self dismissController:self];
	if (![UserData isLoggedIn]) {
		[NSApp terminate:self];
	}
}

//MARK: - Event Handler
- (void)handleLoggedInNotification:(NSNotification *)notification{
    [[NSApplication sharedApplication] updateDropboxUsage];
    [self dismissController:self];
}



@end

File: AppBox/ViewController/DropboxViewController/DropboxViewController.swift
//
//  DropboxViewController.swift
//  AppBox

import AppKit
import AppBoxCore

public final class DropboxViewController: NSViewController {

    public override func loadView() {
        let quitTitle = UserData.isLoggedIn() ? "Cancel" : "Quit"
        view = DropboxLoginHost.makeView(
			quitTitle: quitTitle,
			onConnect: { [weak self] in
				self?.connectDropbox()
			},
			onQuit: { [weak self] in
				self?.quitTapped()
			})
    }

    public override func viewDidLoad() {
        super.viewDidLoad()
        NotificationCenter.default.addObserver(
			self,
			selector: #selector(handleLoggedInNotification(_:)),
			name: Notification.Name("DropBoxLoggedInNotification"),
			object: nil)
    }

    private func connectDropbox() {
        DropboxLogin.present(from: self)
    }

    private func quitTapped() {
        dismiss(self)
        if !UserData.isLoggedIn() {
            NSApp.terminate(self)
        }
    }

    @objc private func handleLoggedInNotification(_ notification: Notification) {
        NSApp.updateAccountsMenu()
        dismiss(self)
    }

    deinit {
        NotificationCenter.default.removeObserver(self)
    }
}
